src/vm/i386/umthunkstub.S

```diff
@@ -154,34 +153,48 @@ LOCAL_LABEL(UMThunkStub_CopyStackArgs):
     and     esp, -16    // align with 16 byte
     lea     edi, [esp]
 
+    // First, we copy arguments to ecx and edx registers (if needed).
     mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
     mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
     mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
     cmp     ebx, -1
-    je      LOCAL_LABEL(CopyStack)
+    je      LOCAL_LABEL(InitCopyStack)
     mov     ecx, dword ptr [esi + ebx]
     add     eax, -4
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
     mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset]
     cmp     ebx, -1
-    je      LOCAL_LABEL(CopyStack)
+    je      LOCAL_LABEL(InitCopyStack)
     mov     edx, dword ptr [esi + ebx]
     add     eax, -4
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
-
-LOCAL_LABEL(CopyStack):
+LOCAL_LABEL(InitCopyStack):
     push    ecx
-    mov     ebx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
-    mov     ebx, dword ptr [ebx + UMEntryThunk__m_pUMThunkMarshInfo]
-    mov     ebx, [ebx + UMThunkMarshInfo__m_cbActualArgSize]
-    sub     ebx, eax
+    push    edx
+    mov     eax, 0  // Offset in edi
+    mov     ebx, 0  // Offset in esi
+LOCAL_LABEL(CopyStack):
+    mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
+    mov     ecx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
+    cmp     ecx, ebx
+    je      LOCAL_LABEL(IncreaseOffset)
+    mov     ecx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset]
+    cmp     ecx, ebx
+    je      LOCAL_LABEL(IncreaseOffset)
     mov     ecx, dword ptr [esi + ebx]
-    mov     dword ptr [edi + ebx], ecx
+    mov     dword ptr [edi + eax], ecx
+    add     eax, 4
+LOCAL_LABEL(IncreaseOffset):
+    add     ebx, 4
+    mov     ecx, [edx + UMThunkMarshInfo__m_cbActualArgSize]
```

Comment:
If you changed the loop to go backwards as it was before, you will not need to access the [edx + UMThunkMarshInfo__m_cbActualArgSize]. 
You would initialize ebx with the actual arg size, use eax that we have before the loop and use [esi + ebx - 4] and [edi + eax - 4 ] above and here you would just terminate when the ebx became zero.
Or, you can just add eax - 4 to the edi before the loop, use [edi] to store to the target and use sub edi, 4 instead of sub eax, 4. Then you could use eax instead of ecx as a helper register in the loop and you would not have to push ecx.